src/main/scala/firrtl/AddDescriptionNodes.scala

```diff
@@ -71,44 +125,85 @@ class AddDescriptionNodes extends Transform {
-  def onStmt(compMap: Map[String, Seq[String]])(stmt: Statement): Statement = {
-    stmt.map(onStmt(compMap)) match {
-      case d: IsDeclaration if compMap.contains(d.name) =>
-        DescribedStmt(DocString(StringLit.unescape(compMap(d.name).mkString("\n\n"))), d)
-      case other => other
+  def onStmt(compMap: Map[String, Seq[Description]])(stmt: Statement): Statement = {
+    val s = stmt.map(onStmt(compMap))
+    val sname = s match {
+      case d: IsDeclaration => Some(d.name)
+      case _ => None
+    }
+    val descs = sname.flatMap({ case name =>
+      compMap.get(name)
+    })
+    (descs, s) match {
+      case (Some(d), DescribedStmt(prevDescs, ss)) => DescribedStmt(prevDescs ++ d, ss)
+      case (Some(d), ss) => DescribedStmt(d, ss)
+      case (None, _) => s
     }
   }
 
-  def onModule(modMap: Map[String, Seq[String]], compMaps: Map[String, Map[String, Seq[String]]])
+  def onModule(modMap: Map[String, Seq[Description]], compMaps: Map[String, Map[String, Seq[Description]]])
     (mod: DefModule): DefModule = {
-    val (newMod, portDesc: Map[String, Description]) = compMaps.get(mod.name) match {
-      case None => (mod, Map.empty)
-      case Some(compMap) => (mod.mapStmt(onStmt(compMap)), mod.ports.collect {
-        case p @ Port(_, name, _, _) if compMap.contains(name) =>
-          name -> DocString(StringLit.unescape(compMap(name).mkString("\n\n")))
-      }.toMap)
-    }
+    val compMap = compMaps.getOrElse(mod.name, Map())
+    val newMod = mod.mapStmt(onStmt(compMap))
+    val portDesc = mod.ports.collect {
+      case p @ Port(_, name, _, _) if compMap.contains(name) =>
+        name -> compMap(name)
+    }.toMap
 
-    val modDesc = modMap.get(newMod.name).map {
-      desc => DocString(StringLit.unescape(desc.mkString("\n\n")))
-    }
+    val modDesc = modMap.get(newMod.name).getOrElse(Seq())
 
     if (portDesc.nonEmpty || modDesc.nonEmpty) {
-      DescribedMod(modDesc.getOrElse(EmptyDescription), portDesc, newMod)
+      DescribedMod(modDesc, portDesc, newMod)
     } else {
       newMod
     }
   }
 
-  def collectMaps(annos: Seq[Annotation]): (Map[String, Seq[String]], Map[String, Map[String, Seq[String]]]) = {
-    val modMap = annos.collect {
-      case DescriptionAnnotation(ModuleName(m, CircuitName(c)), desc) => (m, desc)
-    }.groupBy(_._1).mapValues(_.map(_._2))
+  /**
+    * Merges descriptions of like types.
+    *
+    * Multiple DocStrings on the same object get merged together into one big multi-line comment.
+    * Similarly, multiple attributes on the same object get merged into one attribute with attributes separated by
+    * commas.
+    * @param descs List of `Description`s that are modifying the same object
+    * @return List of `Description`s with some descriptions merged
+    */
+  def mergeDescriptions(descs: Seq[Description]): Seq[Description] = {
+    val (docs: Seq[DocString], nodocs) = descs.partition(_ match {
+      case _: DocString => true
+      case _ => false
+    })
```

Comment:
This is a minor stylistic nit, but `_ match {` is never (? I think) necessary, you can just write:
```scala
    val (docs: Seq[DocString], nodocs) = descs.partition {
      case _: DocString => true
      case _ => false
    }
```
Also I really wish Scala had like a `collectPartition` where you could just provide the `true` case and the `false` would go to the other. Alas.

Reply:
Thanks, this is a good habit I should get into and looks much nicer.